src/main/java/uk/ac/cam/cl/dtg/segue/search/ElasticSearchProvider.java

```diff
@@ -638,44 +641,45 @@
         return result;
     }
 
-    private QueryBuilder processMatchInstructions(final AbstractMatchInstruction matchInstruction)
+
+    /**
+     * Based on the relatively abstract {@code matchInstruction}, generates a {@code QueryBuilder} which is usable by
+     * Elasticsearch.
+     *
+     * @param matchInstruction An {@code AbstractMatchInstruction} representing a search query.
+     *
+     * @return a {@code QueryBuilder} reflecting the instructions in {@code matchInstruction}.
+     * @throws SegueSearchException
+     */
+    private QueryBuilder processMatchInstructions(final AbstractInstruction matchInstruction)
             throws SegueSearchException {
-        if (matchInstruction instanceof BooleanMatchInstruction) {
-            BooleanMatchInstruction booleanMatch = (BooleanMatchInstruction) matchInstruction;
+        if (matchInstruction instanceof BooleanInstruction) {
```

Comment:
To be fair, GitHub has a point here - it would be much cleaner (especially if we are okay with ditching `ISearchProvider`), if the logic for turning each `AbstractInstruction` into a `QueryBuilder` was inside an abstract method `AbstractInstruction::compile` or something, which each subclass implements. That way you wouldn't even need this method, you could just call `matchInstruction.compile()`.

e.g.
```java
public class BooleanInstruction extends AbstractInstruction {
    
    ...
    
    @Override
    public QueryBuilder compile() {
        BoolQueryBuilder query = QueryBuilders.boolQuery();
        for (AbstractInstruction should : booleanMatch.getShoulds()) {
            query.should(should.compile());
        }
        for (AbstractInstruction must : booleanMatch.getMusts()) {
            query.must(must.compile());
        }
        for (AbstractInstruction mustNot : booleanMatch.getMustNots()) {
            query.mustNot(mustNot.compile());
        }
        if (booleanMatch.getBoost() != null) {
            query.boost(this.getBoost());
        }
        query.minimumShouldMatch(this.getMinimumShouldMatch());
        return query;
    }
}
```

I guess it depends whether you want to keep all the code relevant to each `AbstractInstruction` together, or whether you want to fully segregate Elasticsearch code from non-Elasticsearch code.

Reply:
I agree with getting rid of ISearchProvider long-term, since it already feels very ES-specific, but I will leave this for now.